include/dxc/DXIL/DxilFunctionProps.h

```diff
@@ -25,6 +25,93 @@ class Constant;
 } // namespace llvm
 
 namespace hlsl {
+
+// SM 6.6 allows WaveSize specification for only a single required size.
+// SM 6.8+ allows specification of WaveSize as a min, max and preferred value.
+struct DxilWaveSize {
+  unsigned Min = 0;
+  unsigned Max = 0;
+  unsigned Preferred = 0;
+
+  DxilWaveSize() = default;
+  DxilWaveSize(unsigned min, unsigned max = 0, unsigned preferred = 0)
+      : Min(min), Max(max), Preferred(preferred) {}
+  DxilWaveSize(const DxilWaveSize &other) = default;
+  DxilWaveSize &operator=(const DxilWaveSize &other) = default;
+  bool operator==(const DxilWaveSize &other) const {
+    return Min == other.Min && Max == other.Max && Preferred == other.Preferred;
+  }
+
+  // Encode valid DxilWaveSize, handling potential degenerate cases.
+  static DxilWaveSize Encode(unsigned min, unsigned max = 0,
+                             unsigned preferred = 0) {
+    if (max == min)
+      max = 0;
+    if (max == 0 && preferred == min)
+      preferred = 0;
+    return DxilWaveSize(min, max, preferred);
+  }
+
+  // Valid non-zero values are powers of 2 between 4 and 128, inclusive.
+  static bool IsValidValue(unsigned Value) {
+    return (Value >= 4 && Value <= 128 && ((Value & (Value - 1)) == 0));
+  }
+  // Valid encodings:
+  //  0, 0, 0: Not defined
+  //  Min, 0, 0: single WaveSize (SM 6.6/6.7)
+  //  Min, Max (> Min), 0 or Preferred (>= Min and <= Max): Range (SM 6.8+)
+  enum class ValidationResult {
+    Success,
+    InvalidMin,
+    InvalidMax,
+    InvalidPreferred,
+    MaxOrPreferredWhenUndefined,
+    PreferredWhenNoRange,
+    MaxEqualsMin,
+    MaxLessThanMin,
+    PreferredOutOfRange,
+  };
+  ValidationResult Validate() const {
+    if (Min == 0) { // Not defined
```

Comment:
Per the spec, min == 0 should be an error regardless

Reply:
You seem to be interpreting what this struct is for in a very different way than intended.

It is not a definition of the input attribute, nor is it a definition of a single encoding in dxil metadata, PSV0, or RDAT.  It is an internal canonical representation of all states for WaveSize.

The valid states for this representation include the undefined, single wave size, and range forms.  In this internal canonical representation, Min == 0 means no wave size is defined.